File: PredicttheWinner6.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    bool PredictTheWinner(vector<int>& nums) {
        vector<vector<int> > dp(nums.size(), vector<int>(nums.size()));
        for(int i = nums.size() - 1; i >= 0; i--) {
        	for(int j = i; j < nums.size(); j++) {
        		if(i == j) dp[i][j] = nums[i];
        		else dp[i][j] = max(nums[i] - dp[i + 1][j], nums[j] - dp[i][j - 1]);
			}
		}
		
		return dp[0][nums.size() - 1] >= 0;
    }
    
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	Solution *solution = new Solution();
	solution->PredictTheWinner(nums) ? cout<<"true" : cout<<"false";
	
	return 0;
}

File: PredicttheWinner6.py
class Solution:
    def PredictTheWinner(self, nums: list) -> bool:
        n = len(nums)
        dp = [0] * n
        for i in range(n - 1, -1, -1):
            for j in range(i, n, 1):
                if i == j:
                    dp[j] = nums[i]
                else:
                    dp[j] = max(nums[i] - dp[j], nums[j] - dp[j - 1])
        return dp[n - 1] >= 0


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.PredictTheWinner(nums))
